File: swim-system-js/swim-core-js/@swim/codec/main/Tag.ts
// Copyright 2015-2019 SWIM.AI inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import {HashCode} from "@swim/util";
import {Mark} from "./Mark";
import {Span} from "./Span";
import {Output} from "./Output";
import {Display} from "./Display";
import {Debug} from "./Debug";
import {Format} from "./Format";

/**
 * Description of a source location.  Tags are used to annotate input sources,
 * particularly for [[Diagnostic diagnostic]] purposes.  A [[Mark]] tag
 * annotates a source position.  A [[Span]] tag annotate a source range.
 *
 * @see [[Diagnostic]]
 */
export abstract class Tag implements Display, Debug, HashCode {
  /**
   * Returns the first source position covered by this `Tag`.
   */
  abstract start(): Mark;

  /**
   * Returns the last source position covered by this `Tag`.
   */
  abstract end(): Mark;

  /**
   * Returns a `Tag` that includes all source locations covered by
   * both this tag, and `that` tag.
   */
  abstract union(that: Tag): Tag;

  /**
   * Returns the position of this `Tag` relative to the given `mark`.
   */
  abstract shift(mark: Mark): Tag;

  abstract display(output: Output): void;

  abstract debug(output: Output): void;

  abstract equals(that: unknown): boolean;

  abstract hashCode(): number;

  // Forward type declarations
  /** @hidden */
  static Mark: typeof Mark; // defined by Mark
  /** @hidden */
  static Span: typeof Span; // defined by Span
  /** @hidden */
  static Format: typeof Format; // defined by Format
}

File: swim-system-java/swim-core-java/swim.codec/src/main/java/swim/codec/Tag.java
// Copyright 2015-2019 SWIM.AI inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package swim.codec;

/**
 * Description of a source location.  Tags are used to annotate input sources,
 * particularly for {@link Diagnostic diagnostic} purposes.  A {@link Mark} tag
 * annotates a source position.  A {@link Span} tag annotate a source range.
 *
 * @see Diagnostic
 */
public abstract class Tag implements Display, Debug {
  /**
   * Returns the first source position covered by this {@code Tag}.
   */
  public abstract Mark start();

  /**
   * Returns the last source position covered by this {@code Tag}.
   */
  public abstract Mark end();

  /**
   * Returns a {@code Tag} that includes all source locations covered by
   * both this tag, and some {@code other} tag.
   */
  public abstract Tag union(Tag other);

  /**
   * Returns the position of this {@code Tag} relative to the given
   * {@code mark}.
   */
  public abstract Tag shift(Mark mark);
}
